mapedit/studio.cc

```diff
@@ -3252,6 +3343,46 @@ void convertToUTF8::convert(gchar *&_convstr, const char *str, const char *enc)
 		_convstr = g_strdup("");
 		return;
 	}
+	// The ASCII codepage stands for the Original Ultima VII French / German.
+	//   It used a 7 bits ASCII codepage with reused codebytes 01 .. 1f
+	//      except 09 0a 1a 1b for French, German and Spanish characters.
+	bool is_7bit = (strcmp(enc, "ASCII") == 0);
+	char *cvt_str = nullptr;
+	if (is_7bit) {
+		if ((cvt_str = g_strdup(str)) != nullptr) {
+			for (unsigned char *ch = reinterpret_cast<unsigned char *>(cvt_str); *ch; ch++) {
+				// 80 <- 01 Ç:C cedilla,    81 <- 02 ü:u diaeresis,  82 <- 03 é:e acute,
+				// 83 <- 04 â:a circumflex, 84 <- 05 ä:a diaeresis,  85 <- 06 à:a grave.
+				if      (*ch <= 0x06) *ch += 0x7f;
+				// 87 <- 07 ç:c cedilla,    88 <- 08 ê:e circumflex.
+				else if (*ch <= 0x08) *ch += 0x80;
+				else if (*ch <= 0x0a) ;
+				// 89 <- 0b ë:e diaeresis,  8a <- 0c è:e grave.
+				else if (*ch <= 0x0c) *ch += 0x7e;
+				else if (*ch <= 0x0d) ;
+				// 8b <- 0e ï:i diaeresis,  8c <- 0f î:i circumflex, 8d <- 10 ì:i grave,
+				// 8e <- 11 Ä:A diaeresis.
+				else if (*ch <= 0x11) *ch += 0x7d;
+				// 90 <- 12 É:E acute E.
+				else if (*ch <= 0x12) *ch += 0x7e;
+				// 93 <- 13 ô:o circumflex, 94 <- 14 ö:o diaeresis.
+				else if (*ch <= 0x14) *ch += 0x80;
+				// 96 <- 15 û:u circumflex, 97 <- 16 ù:u grave.
+				else if (*ch <= 0x16) *ch += 0x81;
+				// 99 <- 17 Ö:O diaeresis,  9a <- 18 Ü:U diaeresis.
+				else if (*ch <= 0x18) *ch += 0x82;
+				// a0 <- 19 á:a acute.
+				else if (*ch <= 0x19) *ch += 0x87;
+				else if (*ch <= 0x1b) ;
+				// e1 <- 1c ß:s sharp.
+				else if (*ch <= 0x1c) *ch += 0xc5;
+				// a1 <- 1d í:i acute,      a2 <- 1e ó:o acute,      a3 <- 1f ú:u acute.
+				else if (*ch <= 0x1f) *ch += 0x84;
+			}
```

Comment:
Maybe this should also be extracted to a separate function, just to be consistent with the above code blocks being extracted away.

Reply:
I can extract them into separate functions.

However the second occurrence of the first conversion comes from the conversion failure recovery path.
Could you comment on whether that path is still needed ?

On the other hand I could place the whole failure recovery path as an else branch of the conversion success path, and place the byte move at the end of the complete if. Would be there once only then ...